### PR TITLE
Add support for {...,...,...} choice in EditorConfig

### PR DESCRIPTION
**Pull request by agocke**
EditorConfig allows for `{option1,option2}` glob matching where either `option1` or `option2` can be
matched during section parsing. This PR adds support for this feature to the parser.

### CODE REVIEW

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs [hunk 1]
```diff
@@ -82,13 +97,46 @@ public static string TryCompileSectionNameToRegEx(string sectionName)
...
+                        // We need to look ahead to disambiguate. Looking for {num..num}
+                        // is easier because it can't be recursive.
+                        var range = TryParseNumberRange(ref lexer);
+                        if (range is null)
+                        {
+                            // Not a number range. Try a choice expression
+                            if (!TryCompileChoice(ref lexer, sb))
+                            {
+                                return false;
+                            }
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: The implicit fall through here to break and then the method return is a bit hard to follow. Consider moving to an explicit return at this location.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs [hunk 2]
```diff
@@ -97,74 +145,163 @@ public static string TryCompileSectionNameToRegEx(string sectionName)
...
+                }
+                else if (lastChar == '}')
+                {
+                    // Close out the capture group
+                    sb.Append(")");
+                    return true;
+                }
+                else
+                {
+                    throw ExceptionUtilities.UnexpectedValue(lastChar);
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: Why not return false here? This comes from bad user input hence I'd expect just a failed parse here, not an exception.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs [hunk 3]
```diff
@@ -97,74 +145,163 @@ public static string TryCompileSectionNameToRegEx(string sectionName)
...
+                }
+            }
+
+            // Propagate failure
+            return false;
+        }
+
+        private static (string numStart, string numEnd)? TryParseNumberRange(ref SectionNameLexer lexer)
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: Consider a comment here explaining what type of range is being parsed.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs [hunk 4]
```diff
@@ -97,74 +145,163 @@ public static string TryCompileSectionNameToRegEx(string sectionName)
...
+                lexer.EatCurrentCharacter() != '.' || lexer.IsDone)
+            {
+                lexer.Position = saved;
+                return null;
+            }
+
+            // Now another number
+            var num2 = lexer.TryLexNumber();
+            if (num2 is null || lexer.IsDone ||
+                lexer.Lex() != TokenKind.CloseCurly)
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: Seems a bit inconsistent to validate the close curly here but not the open one. Would expect this parsing routine to be completely ignorant of curlies or verify both.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs [hunk 5]
```diff
@@ -179,7 +316,38 @@ public TokenKind Lex()
...
-            public char EatCurrentCharacter() => _sectionName[_position++];
+            public char EatCurrentCharacter() => _sectionName[Position++];
+
+            public char this[int position] => _sectionName[position];
+
+            /// <summary>
+            /// Returns the string representation of a decimal integer, or null if
+            /// the current lexeme is not an integer.
+            /// </summary>
+            public string TryLexNumber()
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: Given we are using `ref struct` here already did you consider defining a `ref struct` to represent a text range here?

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs [hunk 6]
```diff
@@ -179,7 +316,38 @@ public TokenKind Lex()
...
+                var sb = new StringBuilder();
+
+                while (!IsDone)
+                {
+                    char currentChar = EatCurrentCharacter();
+                    if (start && currentChar == '-')
+                    {
+                        sb.Append('-');
+                    }
+                    else if (currentChar >= '0' && currentChar <= '9')
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: Char.IsDigit?

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs [hunk 7]
```diff
@@ -179,7 +316,38 @@ public TokenKind Lex()
...
+            {
+                bool start = true;
+                var sb = new StringBuilder();
+
+                while (!IsDone)
+                {
+                    char currentChar = EatCurrentCharacter();
+                    if (start && currentChar == '-')
+                    {
+                        sb.Append('-');
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: What does a `-` mean in the number representation here?

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs (hunk 2 above)
Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: `TryCompilePathList` should only return true if the `<path-list>` non-terminal was properly parsed, meaning it must have ended in a `,` or a `}`. If it doesn't, `TryCompilePathList` should have returned `false`, so we would never get to this line.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs (hunk 5 above)
Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: Seemed like a tuple was sufficient here, since it's just a pair.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs (hunk 7 above)
Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: Negative numbers. I really dislike this feature.

[the remark this answers is not in the document]
Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: Number ranges aren't supported yet (still prototype).

[the remark this answers is not in the document]
Comment (jaredpar) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: If it's prototype only then lets add a comment stating this code is untested and tests need to be added. 

---
In reply to: [199957075](https://github.com/dotnet/roslyn/pull/28083#discussion_r199957075) [](ancestors = 199957075)

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs (hunk 5 above)
Comment (jaredpar) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: Don't quite understand your response here as there is no tuple. 

---
In reply to: [199956869](https://github.com/dotnet/roslyn/pull/28083#discussion_r199956869) [](ancestors = 199956869)

**Comment by jaredpar** [comment]
done with pass (iteration 2)

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs (hunk 5 above)
Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: Oh, never mind, wrong method. What would be in the ref-struct? Just a string? I didn't think that was necessary.

Comment (jaredpar) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: Basically equivalent of `Span<char>`. It's not strictly necessary here. Given you were using ref struct I watned to know if it's an option you considered.

---
In reply to: [201843061](https://github.com/dotnet/roslyn/pull/28083#discussion_r201843061) [](ancestors = 201843061)

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs [hunk 8]
```diff
@@ -97,74 +155,188 @@ public static string TryCompileSectionNameToRegEx(string sectionName)
...
+            var num1 = lexer.TryLexNumber();
+            if (num1 is null || lexer.IsDone)
+            {
+                // Not a number
+                lexer.Position = saved;
+                return null;
+            }
+
+            // The next two characters must be ".."
+            if (lexer.EatCurrentCharacter() != '.' || lexer.IsDone ||
```
Comment (333fred) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: >EatCurrentCharacter() != '.' || lexer.IsDone [](start = 22, length = 44)

Consider creating a TryEatCurrentCharacter or similar so you don't have to keep checking lexer.IsDone everywhere. Seems fragile the way that it's done now.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs [hunk 9]
```diff
@@ -97,74 +155,188 @@ public static string TryCompileSectionNameToRegEx(string sectionName)
...
+        private static (string numStart, string numEnd)? TryParseNumberRange(ref SectionNameLexer lexer)
+        {
+            var saved = lexer.Position;
+            if (lexer.Lex() != TokenKind.OpenCurly)
+            {
+                lexer.Position = saved;
+                return null;
+            }
+
+            var num1 = lexer.TryLexNumber();
```
Comment (333fred) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: >num1 [](start = 16, length = 4)

Consider aligning these names with the tuple element names.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs [hunk 10]
```diff
@@ -179,7 +351,39 @@ public TokenKind Lex()
...
+                    {
+                        sb.Append(currentChar);
+                    }
+                    else
+                    {
+                        break;
+                    }
+                    start = false;
+                }
+                return sb.Length > 0 ? sb.ToString() : null;
```
Comment (333fred) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.SectionNameMatching.cs]: >sb.Length > 0 [](start = 23, length = 13)

What if all you got was a `-` followed by non-numbers?

File: src/Compilers/Core/CodeAnalysisTest/EditorConfigTests.cs [hunk 11]
```diff
@@ -394,6 +394,133 @@ public void LiteralQuestions()
...
+            Assert.Equal("^.*/[^/]*\\.(?:cs|vb|fs)$", regex);
+
+            Assert.Matches(regex, "/abc.cs");
+            Assert.Matches(regex, "/abc.vb");
+            Assert.Matches(regex, "/abc.fs");
+            Assert.Matches(regex, "/subdir/abc.cs");
+            Assert.Matches(regex, "/subdir/abc.vb");
+            Assert.Matches(regex, "/subdir/abc.fs");
+
+            Assert.DoesNotMatch(regex, "/abcxcs");
```
Comment (333fred) [inline src/Compilers/Core/CodeAnalysisTest/EditorConfigTests.cs]: Consider adding a test with a `.(cs|vb|fs)` in the middle of the string to ensure that we're matching to the end of the string.

File: src/Compilers/Core/CodeAnalysisTest/EditorConfigTests.cs [hunk 12]
```diff
@@ -394,6 +394,133 @@ public void LiteralQuestions()
...
+            // not a section contains a slash, not whether any of the choices contain a slash. So while
+            // [*.cs] usually translates into '**/*.cs' because it contains no slashes, the slashes in
+            // the second choice make this into '/*.cs', effectively matching only files in the root
+            // directory of the match, instead of all subdirectories.
+            Assert.Equal("^/(?:[^/]*\\.cs|subdir/test\\.vb)$", regex);
+
+            Assert.Matches(regex, "/test.cs");
+            Assert.Matches(regex, "/subdir/test.vb");
+
+            Assert.DoesNotMatch(regex, "/subdir/test.cs");
```
Comment (333fred) [inline src/Compilers/Core/CodeAnalysisTest/EditorConfigTests.cs]: Same comment as above.

**Comment by 333fred** [comment]
Done review pass (iteration 3)
